wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/FindId.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.languageserver;
+
+import com.google.common.collect.ImmutableSet;
+import java.util.HashSet;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.languageserver.RegistryContainer.Registry;
+
+/** Utility class that simplifies finding language server id. */
+@Singleton
+class FindId {
+  private final Registry<Set<Pattern>> patterns;
+
+  @Inject
+  FindId(RegistryContainer registryContainer) {
+    this.patterns = registryContainer.patternRegistry;
+  }
+
+  /**
+   * Finds language server ids that corresponds to a specified workspace path
+   *
+   * @param wsPath absolute workspace path
+   * @return set of language server ids
+   */
+  Set<String> byPath(String wsPath) {
+    Set<String> ids = new HashSet<>();
+
+    for (Entry<String, Set<Pattern>> entry : patterns.getAll().entrySet()) {
+      String id = entry.getKey();
+      Set<Pattern> patterns = entry.getValue();
+
+      for (Pattern pattern : patterns) {
+        if (pattern.matcher(wsPath).lookingAt()) {
+          ids.add(id);
+        }
+      }
+    }
+
+    return ImmutableSet.copyOf(ids);
```

Comment:
It is immutable because it is not expected to be changed after it is created. If it is interesting for you there is a nice [link](https://github.com/google/guava/wiki/ImmutableCollectionsExplained).